types/tx.go

```diff
@@ -96,58 +79,83 @@ func (txs Txs) SplitIntoShares() NamespacedShares {
 
 // TxProof represents a Merkle proof of the presence of a transaction in the Merkle tree.
 type TxProof struct {
-	RootHash tmbytes.HexBytes `json:"root_hash"`
-	Data     Tx               `json:"data"`
-	Proof    merkle.Proof     `json:"proof"`
-}
-
-// Leaf returns the hash(tx), which is the leaf in the merkle tree which this proof refers to.
-func (tp TxProof) Leaf() []byte {
-	return tp.Data.Hash()
+	RowRoots []tmbytes.HexBytes  `json:"root_hash"`
+	Data     [][]byte            `json:"data"`
+	Proofs   []*tmproto.NMTProof `json:"proof"`
 }
 
 // Validate verifies the proof. It returns nil if the RootHash matches the dataHash argument,
 // and if the proof is internally consistent. Otherwise, it returns a sensible error.
-func (tp TxProof) Validate(dataHash []byte) error {
-	if !bytes.Equal(dataHash, tp.RootHash) {
-		return errors.New("proof matches different data hash")
-	}
-	if tp.Proof.Index < 0 {
-		return errors.New("proof index cannot be negative")
-	}
-	if tp.Proof.Total <= 0 {
-		return errors.New("proof total must be positive")
-	}
-	valid := tp.Proof.Verify(tp.RootHash, tp.Leaf())
-	if valid != nil {
-		return errors.New("proof is not internally consistent")
+func (tp TxProof) Validate() error {
+	if len(tp.RowRoots) != len(tp.Proofs) || len(tp.Data) != len(tp.Proofs) {
+		return errors.New(
+			"invalid number of proofs, row roots, or data. they all must be the same to verify the proof",
+		)
+	}
+	for _, proof := range tp.Proofs {
+		if proof.Start < 0 {
+			return errors.New("proof index cannot be negative")
+		}
+		if (proof.End - proof.Start) <= 0 {
+			return errors.New("proof total must be positive")
+		}
+		valid := tp.VerifyProof()
+		if !valid {
+			return errors.New("proof is not internally consistent")
+		}
 	}
+
 	return nil
 }
 
-func (tp TxProof) ToProto() tmproto.TxProof {
+func (tp *TxProof) VerifyProof() bool {
```

Comment:
will change this to take a slice of row roots as args, and not use the roots already stored in the `TxProof`